### DISCUSSION TITLE
Anti-cheat issuance of items.

### DISCUSSION THREAD

**Discussion by exrex90** [Pryaxis/TShock#1935] [Ideas]
Hello, add an anti-cheat that will block the issuance of items?

**Comment by moisterrific** [comment]
What do you mean issuance of items? As in cheaters spawning any item into their inventory or duplication from research in journey mode?

**Reply by exrex90** [reply]
These two problems. But for the first time, scammers have been preventing us from playing normally for 5-6 years (at the moment cheat  a lot for terraria and almost everyone is playing with them, in this case there is no point in SCC mode at all.

**Reply by moisterrific** [reply]
Word from more knowledgeable people than myself is that fundamentally because of how this game is designed, it’s pretty much impossible to prevent cheaters from spawning in any item they want. I’ve seen and tested a few plugins that aim to detect and prevent item spawning but they all have too many false positives to be usable on a public server.

This is definitely a major issue, I’m sure the TShock devs knew about it, but just can’t do anything to patch it because it’s virtually impossible to fix.

**Reply by AceofSpades5757** [reply]
Anything even close? I've been checking inventories and was thinking of at least creating flags for users with obvious signs, such as 999 platinum coins, Hardmode items in Pre-Hardmode, and boss drops for bosses that haven't spawned.

**Reply by exrex90** [reply]
I think it can be done but a lot of work. As I imagine it, we check the world around us and if an object appears but it isn’t near, we delete it, we look at the craft item if we have enough reagents to create it, then we give the item and check that the items for the craft were lost.

**Reply by hakusaro** [reply]
> Word from more knowledgeable people than myself is that fundamentally because of how this game is designed, it’s pretty much impossible to prevent cheaters from spawning in any item they want. I’ve seen and tested a few plugins that aim to detect and prevent item spawning but they all have too many false positives to be usable on a public server.

In SSC it is actually possible to do this, but SSC checks aren't perfect right now.

**Comment by AceofSpades5757** [comment]
I've been running into this problem ever since I started my server a couple of weeks ago. They're rampant.

**Reply by exrex90** [reply]
it is a pain

**Reply by akku9164** [reply]
hello sir can u develop anti hack scrip and monitoring software

**Reply by AceofSpades5757** [reply]
I mean, that's part of we're discussing. I'm hoping there is at least something already out there, especially since Tshock plugins and almost all related software seem to be written in C# while I primarily use Python. The script I wrote for the inventory check is just a Python script to check the SQLite database (SSC) and match item IDs so that I can actually read it.

**Comment by hakusaro** [comment]
We're currently working on many anti-cheat features. Currently, though, we're trying to solve some game breaking issues before we do that. It is helpful if you know precisely which attacks people are doing so that we can take care of them. Currently, I'm aware of at least one method that seems to work to sneak items into a server.

I would say in general that your best defense against this, right now, is to **permanently ban** anyone discovered to have illegitimate items on sight. Do not give these people second chances. Do not let them back onto your server. Even if we patch the item spawning issues, you will still never want to play with these people. They are scum.

**Reply by AceofSpades5757** [reply]
Well you guys are doing some spectacular work and it's understandable with the huge, new update to Terraria.

One guy admitted to using TerraHax and people are clearly just sitting there AFK while creating items in their inventory. Not sure how they do it, but I hope this helps.

**Reply by exrex90** [reply]
We will wait from your team for anti-cheating in Tshock because only you can do this.

**Comment by bartico6** [comment]
tShock will most likely never provide a full server-side anti-cheat for the game.
This is because such an anti-cheat not only requires a lot of computationally expensive logic to be ran on the server, it also requires a lot of maintenance and planning as well as foreseeing every potential use-case people might want from a tShock server.

Therefore, **preventing inventory item cheats is in the plugin scope**, in my opinion, and I personally (*emphasis on the fact I don't speak on behalf of the team now*) don't expect tShock to provide such functionality in the foreseeable future.

**Reply by exrex90** [reply]
Do you have such a plugin?

**Reply by bartico6** [reply]
Yes, but it is not public and I'm not willing to share it.

**Reply by exrex90** [reply]
Сan sell?

**Reply by bartico6** [reply]
No.
